### DISCUSSION TITLE
Explosions below sea level

### DISCUSSION THREAD

**Discussion by FrankHeijden** [GriefPrevention/GriefPrevention#1187] [Q&A]
Hey, I stumbled upon this piece of code in GP: https://github.com/TechFortress/GriefPrevention/blob/master/src/main/java/me/ryanhamshire/GriefPrevention/EntityEventHandler.java#L379

Which raised the question to me why we subtract the arbitrary number 7 from the sea level (which is a value we can configure from the config?)

**Comment by FrankHeijden** [comment]
It seems it was added in a refactor of that method a few commits after that: https://github.com/TechFortress/GriefPrevention/commit/e91924eb94e55acfa1276a9d5ca67987f6b639c3 ; it may be to prevent flying tnt-dupers from destroying unclaimed land?

We might be able to handle such cases differently, a way I can think of is to only allow the explosion to happen if all blocks are below the sea level -- this so we can prevent an edge case where the explosion happens on the boundary of the sea level, where the top half would be "untouched" and the bottom half would be removed by tnt.

We could also be to introduce a percentage of blocks which are below sea level -- for example, if 80% of the blocks in the blocklist are below sea level, allow the explosion to happen.

Another way might be to track the tnt entity and only if the Primed TNT entity is below sea level, allow the explosion to happen.

This last one I think is best, since it is quite intuitive to the user.

**Reply by Jikoo** [reply]
I don't see how blocking half an explosion is an edge case. If anything, due to the unnatural nature of it, it makes it clearer to the user that they're doing something "wrong" and are being stopped.

I think checking the location of the explosion source is a good idea for sure. High yield explosions above surface level could cause inaccessible pockets for mob spawning.

Flat allowing explosions from entities beneath sea level is fine by me as long as it also removes blocks at or above sea level because of higher yield explosions.

**Reply by Jikoo** [reply]
To clarify: I strongly disagree with not removing blocks from the explosion block list based on Y level because of compatibility with plugins adding larger explosions. I do think it would be a very good idea to only allow the explosion if the source is also below sea level. I agree that the arbitrary 7 is confusing, and I think it should be fine to reduce to 1 to make it consistent with what more advanced users might expect based on the message when placing TNT.

**Reply by FrankHeijden** [reply]
This clarifies the confusion from below -- but if other plugins create bigger explosions -- shouldn't those plugins be handling how damage is dealt (and what to do with it)?

**Comment by Muspah** [comment]
Yeah, I'm always struggling when searching in the log locally and on Github at the same time, was referencing the wrong commit. Not quick enough to delete the comment unfortunately <3

**Comment by Jikoo** [comment]
Pretty sure this is a compromise between configurability and duplicate nodes - sea level is used in other places. Exploding directly up to sea level could lead to holes in terrain, so BigScary probably decided it was easier on server admins to not have to figure out an extra few nodes and made it a fixed distance based on the number of low resistance blocks TNT can bust.

**Reply by FrankHeijden** [reply]
The default message sent to the end-user might be a bit confusing to them then: https://github.com/TechFortress/GriefPrevention/blob/master/src/main/java/me/ryanhamshire/GriefPrevention/DataStore.java#L1675 ; this suggests the blocks will explode below sea level (y=62).

We could then make a seperate node in the configuration to split this "sea level" from the tnt level? I am also a fan of not checking per block if it is below sea level, but just based on the tnt entity.

**Reply by Jikoo** [reply]
Per my comment in reply to your other comment: Flat allowing explosions from entities beneath sea level is fine by me as long as it also removes blocks at or above sea level because of higher yield explosions.

TNT is a power of 4 if I remember correctly. What if I explode something with power 10, 20? We need to check block levels for compatibility with other plugins.

**Reply by FrankHeijden** [reply]
> as long as it also removes blocks at or above sea level

Do you mean by this "remove them from the blocklist the tnt is supposed to explode" (i.e. from the event), or "allow the tnt to remove blocks above sea level"? Because in my opinion, I think it is OK to just allow the whole explosion to happen (ie not remove any blocks from the event) if the exploding entity is below sea level -- I don't think you can create such big block explosions (from minecraft wiki a factor of 8 is the max obtained from a wither: https://minecraft.gamepedia.com/Explosion)

**Reply by Jikoo** [reply]
Remove from the event blocklist.

>compatibility with other plugins

We're not just working with vanilla. Why arbitrarily remove support for other plugins when we can keep it easily? If we were including the arbitrary 7 I'd be tempted to agree with you that it's reasonably safe, but to remove both the arbitrary 7 block below sea level limit AND allow explosions to be initiated entirely unchecked directly below sea level is unacceptable to me. Even regular TNT would be able to create surface holes using those rules, which is exactly what this feature is supposed to prevent.
